File: Python/Programmers/KAKAO/문자열 압축.py
# 문자열 s를 1개, 2개, 3개, ... 단위로 압축 즉, 완전탐색 알고리즘 적용

# def solution(s):
#     answer = 1000 # 문자열의 최대 길이는 1000이므로 초기값 1000 설정
#     for i in range(1, len(s) + 1): # 1부터 len(s)까지 압축
#         result = ""
#         check = []
#         cnt = 1
#         for j in range(0, len(s), i): # i만큼 압축 되어야 하므로 인덱스 증가 크기를 i로 설정
#             if (j + i) <= len(s): # (j + i)는 s의 길이보다 작거나 같아야함
#                 tmp = s[j:j+i]
#             else: # s의 길이보다 크다면 tmp에 나머지 문자 저장
#                 tmp = s[j:]

#             if check: # check가 비어있지 않다면
#                 if check[-1] == tmp: # check의 마지막 값이 tmp랑 같다면
#                     cnt += 1 # count
#                     continue

#                 cnt = "" if cnt == 1 else str(cnt) # 같지 않다면 cnt가 1인 것은 생략하므로 1일 때는 빈 문자열, 1이 아닐 때는 cnt를 문자열로 변환
#                 result += (cnt + check[-1]) # result에 압축 결과를 더해나감
#                 cnt = 1 # cnt 값 초기화

#             check.append(tmp) # check에 tmp append
            
#         # 문자열을 압축하고 난 후 마지막 문자를 result에 더함
#         cnt = "" if cnt == 1 else str(cnt)
#         result += (cnt + tmp)

#         if answer > len(result): # 압축 길이가 제일 작은 것을 answer에 저장
#             answer = len(result)

#     return answer

# s의 길이의 절반이 최대 압축할 수 있는 범위, 따라서 s의 길이의 절반을 넘어가게 되면 압축을 할 수 없음
# 따라서 압축 범위는 s 길이의 절반으로 좁힘

def solution(s):
    answer = len(s) # 최대길이가 s의 길이
    for i in range(1, len(s) // 2 + 1): # 압축 범위를 s의 절반으로 좁힘
        result = ""
        check = s[0:i] # 비교할 check 변수의 초기값 설정
        cnt = 1
        for j in range(i, len(s), i):
            if check == s[j:j+i]: # split한 문자열과 check가 같다면 압축할 수 있으므로 count
                cnt += 1
            
            else: # 같지 않다면 압축할 수 없으므로
                result += (str(cnt) + check if cnt > 1 else check) # result에 압축 결과 문자열을 더해나감
                check = s[j:j+i] # check 변수 갱신
                cnt = 1 # count 변수 초기화

        result += (str(cnt) + check if cnt > 1 else check) # 전부 압축하고 마지막 문자열 result에 더함
            
        if answer > len(result):
            answer = len(result)
        
    return answer

File: JavaScript/Programmers/KAKAO/문자열 압축.js
// 1부터 s.length / 2까지 문자열을 압축시킨 후 제일 짧은 길이 반환
// s.length / 2인 이유는 s.length / 2보다 클 경우에는 압축될 수 없으므로 s.length / 2까지만 압축하고 비교
// function solution(s) {
//     let answer = s.length; // 최대값은 s.length이므로 초기값 설정
//     for (let step = 1; step <= s.length / 2; step++) { // s.length / 2까지만 압축하는게 의미 있으므로 s.length / 2까지만 순회
//         const zipped = []; // 압축 결과들을 담은 배열
//         for (let index = 0; index < s.length; index += step) { // 문자열을 압축하기 위해 0부터 step씩 증가하면서 순회
//             const slice = s.slice(index, index + step); // 압축 문자열 저장
//             const last = zipped[zipped.length - 1]; // zipped 배열의 마지막 값 가져옴
//             if (last && last[0] === slice) { // last 값이 있고, last의 압축 문자열이 slice와 같다면
//                 zipped[zipped.length - 1] = [last[0], last[1] + 1]; // 마자막 값에 count 값만 + 1 진행
//                 continue;
//             }
//             zipped.push([slice, 1]); // last 값이 없거나, slice와 같지 않다면 push
//         }
//         // zipped 배열 순회하면서 문자열 압축
//         const result = zipped.map(([slice, count]) => `${count === 1 ? "" : count}${slice}`).join("");
//         // 최소값 저장
//         answer = Math.min(answer, result.length);
//     }
    
//     return answer;
// }

function solution(s) {
    let answer = s.length;
    for (let step = 1; step <= s.length / 2; step++) {
        let result = "", tmp = s.slice(0, step), count = 1; // 배열에 담아서 비교하지 않고 그때그때마다 문자열 압축하고 저장함
        for (let index = step; index < s.length; index += step) {
            const slice = s.slice(index, index + step);
            if (tmp === slice) { // tmp 값과 slice 값이 같으면 카운팅
                count += 1;
                continue;
            }
            // tmp 값과 slice 값이 같지 않으면
            // 문자열 압축
            result += `${count === 1 ? "" : count}${tmp}`;
            // tmp, count 값 초기화
            tmp = slice;
            count = 1;
        }
        // for문을 다 순회하고 마지막 압축된 문자열 result에 저장
        result += `${count === 1 ? "" : count}${tmp}`;
        // 최소값 저장
        answer = Math.min(answer, result.length);
    }
    
    return answer;
}
